src/backend/catalog/gp_matview_aux.c

```diff
@@ -0,0 +1,444 @@
+/*-------------------------------------------------------------------------
+ *
+ * gp_matview_aux.c
+ *
+ * Portions Copyright (c) 2024-Present HashData, Inc. or its affiliates.
+ *
+ *
+ * IDENTIFICATION
+ *	    src/backend/catalog/gp_matview_aux.c
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+#include "nodes/pg_list.h"
+#include "nodes/parsenodes.h"
+#include "access/htup.h"
+#include "access/htup_details.h"
+#include "access/table.h"
+#include "access/genam.h"
+#include "catalog/dependency.h"
+#include "catalog/gp_matview_aux.h"
+#include "catalog/gp_matview_tables.h"
+#include "catalog/pg_type.h"
+#include "catalog/indexing.h"
+#include "cdb/cdbvars.h"
+#include "utils/array.h"
+#include "utils/builtins.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
+#include "utils/syscache.h"
+#include "utils/lsyscache.h"
+#include "storage/lockdefs.h"
+#include "optimizer/optimizer.h"
+#include "parser/parsetree.h"
+
+static void InsertMatviewTablesEntries(Oid mvoid, List *relids);
+
+static void RemoveMatviewTablesEntries(Oid mvoid);
+
+static void SetMatviewAuxStatus_guts(Oid mvoid, char status);
+
+/*
+ * GetViewBaseRelids
+ * Get all base tables's oid of a query tree.
+ * Currently there is only one base table, but there should be
+ * distinct func on it later. Self join tables: t1 join t1, will
+ * get only one oid.
+ * 
+ * Return NIL if the query we think it's useless.
+ */
+List*
+GetViewBaseRelids(const Query *viewQuery)
+{
+	List	*relids = NIL;
+	Node	*mvjtnode;
+
+	if ((viewQuery->commandType != CMD_SELECT) ||
+		(viewQuery->rowMarks != NIL) ||
+		(viewQuery->distinctClause != NIL) ||
+		(viewQuery->scatterClause != NIL) ||
+		(viewQuery->cteList != NIL) ||
+		(viewQuery->groupingSets != NIL) ||
+		(viewQuery->havingQual != NULL) ||
+		(viewQuery->setOperations != NULL) ||
+		viewQuery->hasWindowFuncs ||
+		viewQuery->hasDistinctOn ||
+		viewQuery->hasModifyingCTE ||
+		viewQuery->groupDistinct ||
+		(viewQuery->parentStmtType == PARENTSTMTTYPE_REFRESH_MATVIEW) ||
+		viewQuery->hasSubLinks)
+	{
+		return NIL;
+	}
+
+	/* As we will use views, make it strict to unmutable. */
+	if (contain_mutable_functions((Node*)viewQuery))
+		return NIL;
+
+	if (list_length(viewQuery->jointree->fromlist) != 1)
+		return NIL;
+
+	mvjtnode = (Node *) linitial(viewQuery->jointree->fromlist);
+	if (!IsA(mvjtnode, RangeTblRef))
+		return NIL;
+
+	RangeTblEntry	*rte = rt_fetch(1, viewQuery->rtable);
+	if (rte->rtekind != RTE_RELATION)
+		return NIL;
+
+	/* Only support normal relation now. */
+	if (get_rel_relkind(rte->relid) != RELKIND_RELATION)
+		return NIL;
+
+	relids = list_make1_oid(rte->relid);
```

Comment:
We can only process one Range Table Entry?

Reply:
Yeah, for now. 
AQUMV and Append Agg plan only could use one base table now.
We could store more complex query if we are ready to use them, ex: a join.

Reply:
Sounds good.